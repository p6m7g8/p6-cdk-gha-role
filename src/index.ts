export * from './p6cdkgharole'
