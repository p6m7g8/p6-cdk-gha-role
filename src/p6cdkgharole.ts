import type { Construct } from 'constructs'
import * as cdk from 'aws-cdk-lib'
import * as iam from 'aws-cdk-lib/aws-iam'

export interface IP6CDKGHARoleProps {
  readonly principle: cdk.Arn
  readonly repo: string
  readonly policies?: iam.IManagedPolicy[]
}

export class P6CDKGHARole extends cdk.Resource {
  constructor(scope: Construct, id: string, props: IP6CDKGHARoleProps) {
    super(scope, id)

    const role = new iam.Role(this, 'gha-p6m7g8-p6-roles', {
      assumedBy: new iam.FederatedPrincipal(
        props.principle as unknown as string,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            [`token.actions.githubusercontent.com:sub`]: `repo:${props.repo}:*`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    })
    if (props.policies) {
      props.policies.forEach(policy => role.addManagedPolicy(policy))
    }
  }
}
