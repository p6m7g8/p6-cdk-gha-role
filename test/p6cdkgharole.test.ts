import * as cdk from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import * as iam from 'aws-cdk-lib/aws-iam'
import { P6CDKGHARole } from '../src'

it('p6CDKWebsitePlus snapshot test', () => {
  const app = new cdk.App()
  const stack = new cdk.Stack(app, 'MyStack')

  // WHEN
  new P6CDKGHARole(stack, 'p6-cdk-name', {
    principle: 'arn:aws:iam::346733622331:oidc-provider/token.actions.githubusercontent.com',
    repo: 'p6m7g8/p6m7g8.com',
    policies: [
      iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'),
      iam.ManagedPolicy.fromAwsManagedPolicyName('CloudFrontFullAccess'),
    ],
  })

  // THEN
  const template = Template.fromStack(stack).toJSON()
  expect(template).toMatchSnapshot()
})
