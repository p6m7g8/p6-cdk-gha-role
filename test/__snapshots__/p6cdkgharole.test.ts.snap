// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6CDKWebsitePlus snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "p6cdknameghap6m7g8p6roles38BE452E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:p6m7g8/p6m7g8.com:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "arn:aws:iam::346733622331:oidc-provider/token.actions.githubusercontent.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonS3FullAccess",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudFrontFullAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
